File: src/main/java/com/timgroup/gradle/webpack/MochaTestTask.java
package com.timgroup.gradle.webpack;

import java.io.File;
import java.util.Arrays;
import javax.inject.Inject;

import org.gradle.api.DefaultTask;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.InputDirectory;
import org.gradle.api.tasks.InputFile;
import org.gradle.api.tasks.OutputFile;
import org.gradle.api.tasks.SkipWhenEmpty;
import org.gradle.api.tasks.TaskAction;
import org.gradle.api.tasks.VerificationTask;
import org.gradle.process.internal.ExecAction;
import org.gradle.process.internal.ExecActionFactory;

public class MochaTestTask extends DefaultTask implements VerificationTask {
    private File mainFiles;
    private File testFiles;
    private File testOutput;
    private File mochaOptionsFile;
    private boolean ignoreFailures;

    @InputDirectory
    public File getMainFiles() {
        return mainFiles;
    }

    public void setMainFiles(Object mainFiles) {
        this.mainFiles = getProject().file(mainFiles);
    }

    @SkipWhenEmpty
    @InputDirectory
    public File getTestFiles() {
        return testFiles;
    }

    public void setTestFiles(Object testFiles) {
        this.testFiles = getProject().file(testFiles);
    }

    @InputFile
    public File getMochaOptionsFile() {
        return mochaOptionsFile;
    }

    public void setMochaOptionsFile(Object mochaOptionsFile) {
        this.mochaOptionsFile = getProject().file(mochaOptionsFile);
    }

    @OutputFile
    public File getTestOutput() {
        return testOutput;
    }

    public void setTestOutput(Object testOutput) {
        this.testOutput = getProject().file(testOutput);
    }

    @Override
    @Input
    public boolean getIgnoreFailures() {
        return ignoreFailures;
    }

    @Override
    public void setIgnoreFailures(boolean ignoreFailures) {
        this.ignoreFailures = ignoreFailures;
    }

    @Inject
    protected ExecActionFactory getExecActionFactory() {
        throw new UnsupportedOperationException();
    }

    @TaskAction
    public void runTests() {
        ExecAction execAction = getExecActionFactory().newExecAction();
        execAction.setExecutable("node_modules/mocha/bin/mocha");
        execAction.environment("JUNIT_REPORT_PATH", testOutput.toString());
        execAction.environment("JUNIT_REPORT_STACK", "1");
        execAction.setArgs(Arrays.asList("--reporter", "mocha-jenkins-reporter", "--opts", mochaOptionsFile, "--recursive", testFiles));
        execAction.setIgnoreExitValue(ignoreFailures);
        execAction.execute();
    }
}

File: src/main/kotlin/com/timgroup/gradle/webpack/MochaTestTask.kt
package com.timgroup.gradle.webpack

import com.moowork.gradle.node.exec.NodeExecRunner
import org.gradle.api.DefaultTask
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputDirectory
import org.gradle.api.tasks.InputFile
import org.gradle.api.tasks.Internal
import org.gradle.api.tasks.OutputFile
import org.gradle.api.tasks.SkipWhenEmpty
import org.gradle.api.tasks.TaskAction
import org.gradle.process.ExecResult
import java.io.File

open class MochaTestTask : DefaultTask() {
    @get:InputDirectory
    var mainFiles: File? = null
    @get:InputDirectory
    @get:SkipWhenEmpty
    var testFiles: File? = null
    @get:InputFile
    var mochaOptionsFile: File? = null
    @get:OutputFile
    var testOutput: File? = null
    @get:Internal
    var result: ExecResult? = null
    @get:Input
    var ignoreFailures: Boolean = false

    @TaskAction
    fun runTests() {
        val runner = NodeExecRunner(project).apply {
            arguments = listOf("node_modules/mocha/bin/mocha",
                    "--reporter=mocha-jenkins-reporter",
                    "--opts",
                    mochaOptionsFile.toString(),
                    "--recursive",
                    testFiles.toString())

            putenv("JUNIT_REPORT_PATH", testOutput)
            putenv("JUNIT_REPORT_STACK", "1")

            ignoreExitValue = ignoreFailures
        }

        result = runner.execute()
    }
}
